keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md

```diff
@@ -0,0 +1,625 @@
+    ...
+
+     // Scope specifies the scope of this rule. If unespecified, the scope is
+     // not limited.
+     //
+     // +optional
+     Scope ScopeType `json:"scope,omitempty" protobuf:"bytes,3,opt,name=scope"`
+}
+```
+
+### http for localhost
+
+Admission Webhook requires https for all connections. This is not necessary when the webhook in running on the same machine as apiserver (like a side-car or static pod). By relaxing this restriction, it would be possible to run side-cars or static pods on the same machine without cert management. The API change for this feature is documentation and relaxing validation.
+
+Note that there must be caution around any validation relaxing in regard to rolling-back the cluster. It must be documented to the end user not to use the feature until they are not going to roll-back the cluster to a version that does not have this feature. If they do, they need to delete or update these webhook configuration as the old cluster does not allow running insecure webhooks as sidecar. To delete these configurations in an old API server, one must first fix the http validation problem otherwise finalizers would not allow deletion the object.
+
+### timeout configuration
+
+Admission Webhook has a default timeout of 30 seconds today but long running webhooks would result in a poor performance. By adding a timeout field to the configuration, the webhook author can limit the running time of the webhook that either result in failing the API call earlier or ignore the webhook call based on the failure policy. This feature, however, would not let the timeout to be extended more than 30 seconds and the timeout would be defaulted to a shorter value (e.g. 10 seconds) for v1 API while stays 30 second for v1beta API to keep backward compatibility.
+
+```golang
+type Webhook struct {
+    ...
+     // TimeoutSeconds specifies the timeout for this webhook. After the timeout passes,
+     // the webhook call will be ignored or the API call will fail based on the
+     // failure policy.
+     // The timeout value must be between 1 and 30 seconds.
+     // Default to 10 seconds for v1 API and 30 seconds for v1beta1 API.
+     // +optional
+     TimeoutSeconds *int32 `json:"timeoutSeconds,omitempty" protobuf:"varint,8,opt,name=timeoutSeconds"`
+}
+```
+
+### Port configuration
+
+Today Admission Webhook port is always expected to be 443 on service reference. But this limitation was arbitrary and there are cases that Admission Webhook cannot use this port. This feature will add a port field to service based webhooks and allows specifying a port other than 443 for service based webhooks. Specifying port should already be available for URL based webhooks.
+
+```golang
+type ServiceReference struct {
+    ...
+
+     // If specified, the port on the service that hosting webhook.
+     // Default to 443 for backward compatibility.
+     // +optional
+     Port *int32 `json:"port,omitempty" protobuf:"varint,4,opt,name=port"`
```

Comment:
what is the difference between zero and nil?

Reply:
should 0 be invalid?